src/Lean/Widget/InteractiveGoal.lean

```diff
@@ -8,59 +8,70 @@ import Lean.Meta.PPGoal
 
-/-- In the infoview, if multiple hypotheses `h₁`, `h₂` have the same type `α`, they are rendered as `h₁ h₂ : α`.
-We call this a 'hypothesis bundle'. -/
+/-- In the infoview, if multiple hypotheses `h₁`, `h₂` have the same type `α`, they are rendered
+as `h₁ h₂ : α`. We call this a 'hypothesis bundle'. We use `none` instead of `some false` for
+booleans to save space in the json encoding. -/
 structure InteractiveHypothesisBundle where
-  /-- The user-friendly name for each hypothesis.
-  If anonymous then the name is inaccessible and hidden. -/
+  /-- The user-friendly name for each hypothesis. -/
   names : Array Name
   /-- The ids for each variable. Should have the same length as `names`. -/
   fvarIds : Array FVarId
   type : CodeWithInfos
   /-- The value, in the case the hypothesis is a `let`-binder. -/
   val? : Option CodeWithInfos := none
   /-- The hypothesis is a typeclass instance. -/
-  isInstance : Bool
+  isInstance? : Option Bool := none
   /-- The hypothesis is a type. -/
-  isType : Bool
+  isType? : Option Bool := none
   /-- If true, the hypothesis was not present on the previous tactic state.
-      Uses `none` instead of `some false` to save space in the json encoding. -/
+  Only present in tactic-mode goals. -/
   isInserted? : Option Bool := none
   /-- If true, the hypothesis will be removed in the next tactic state.
-      Uses `none` instead of `some false` to save space in the json encoding.  -/
-  isRemoved?  : Option Bool := none
+  Only present in tactic-mode goals. -/
+  isRemoved? : Option Bool := none
   deriving Inhabited, RpcEncodable
 
-structure InteractiveGoal where
-  hyps      : Array InteractiveHypothesisBundle
-  type      : CodeWithInfos
+/-- The shared parts of interactive term-mode and tactic-mode goals. -/
+structure InteractiveGoalCore where
+  hyps : Array InteractiveHypothesisBundle
+  /-- The target type. -/
+  type : CodeWithInfos
+  /-- Metavariable context that the goal is well-typed in. -/
+  ctx : WithRpcRef Elab.ContextInfo
+  deriving Inhabited
+
+/-- An interactive tactic-mode goal. -/
+structure InteractiveGoal extends InteractiveGoalCore where
+  /-- The name `foo` in `case foo`, if any. -/
   userName? : Option String
+  /-- The symbol to display before the target type. Usually `⊢ ` but `conv` goals use `∣ `
+  and it could be extended. -/
   goalPrefix : String
-  /-- Identifies the goal (ie with the unique
-  name of the MVar that it is a goal for.)
-  This is none when we are showing a term goal. -/
-  mvarId? : Option MVarId := none
-  /-- If true, the goal was not present on the previous tactic state.
-      Uses `none` instead of `some false` to save space in the json encoding.  -/
-  isInserted?: Option Bool := none
-  /-- If true, the goal will be removed on the next tactic state.
-      Uses `none` instead of `some false` to save space in the json encoding. -/
+  /-- Identifies the goal (ie with the unique name of the MVar that it is a goal for.) -/
+  mvarId : MVarId
```

Comment:
It is worth discussing what data, which eventually gets passed to user widgets, should be included here. It might also be worthwhile to put the entire `TacticInfo` in `InteractiveGoals` since that contains the `Syntax` which self-replacing tactics that come with associated widgets could use to accurately edit the source. Note that this is independent of code actions which use a different mechanism.

Reply:
Passing more infos to the editor seems like a good idea in general.

Reply:
Ok! After all I am not sure it is needed yet but we can always add it later.